### ISSUE TITLE
WebExtension Issue

### ISSUE THREAD

**Issue by Mandarancio** [fabiocolacio/Marker#74]
In answer  to the comments of #66
 
I tried with this md file [syntax.text](https://daringfireball.net/projects/markdown/syntax.text) and I do not have any problem or glitch on my laptop with the extension. 

Which distribution are you using? What are the version of your webkit2gtk and gtk libraries?

**Comment by fabiocolacio** [comment]
I am using Fedora 27, with WebKit2GTK+ 2.18.4 (installed from the fedora repositories).

**Comment by Mandarancio** [comment]
mmm it's the same that mine on Archlinux... I will try the autoscroll on another computer to be understand better.

**Comment by Mandarancio** [comment]
This is the normal behaviour on my configuration:
![marker_autoscroll](https://user-images.githubusercontent.com/276686/34672541-22bc7d64-f47f-11e7-99e1-7da581c147d1.gif)

**Comment by fabiocolacio** [comment]
I still have the behavior from before the webextensions were added most of the time. When I type a single character it scrolls all the way to the top:

![peek 2018-01-08 13-51](https://user-images.githubusercontent.com/5248352/34694234-315a9f94-f47b-11e7-92b9-67d861d87796.gif)

The web extension works sometimes for me, but very rarely.

I will try on other machines to see how it works there.

**Comment by fabiocolacio** [comment]
When adding print statements into the webextension, the print statements are being executed, so the webextension is being loaded (but I still have the issue above). It must be an implementation issue then. I will do some more tests and report back here.

**Comment by Mandarancio** [comment]
That is very strange... It never happen to me, let me know when you discover the origin of the problem.

**Comment by fabiocolacio** [comment]
I changed the store and restore functions in the web extension as follows:

```C
static void
restore_scroll_position(WebKitWebPage   *web_page,
                        gpointer         user_data)
{
    WebKitDOMDocument * document = webkit_web_page_get_dom_document (web_page);
    WebKitDOMElement* body = WEBKIT_DOM_ELEMENT(webkit_dom_document_get_body (document));
    
    const glong *pos = user_data;
    
    if (body){
        webkit_dom_element_set_scroll_top(body, *pos);     
    } else {
      g_print("couldn't get body!\n");
    }
    
    g_print("restore_scroll_position()\n");
    g_print("  scroll position: %d\n", *pos);
}

static gboolean 
store_scroll_position(WebKitWebPage      *web_page,
                       WebKitURIRequest  *request,
                       WebKitURIResponse *redirected_response,
                       gpointer           user_data)
{
    WebKitDOMDocument * document = webkit_web_page_get_dom_document (web_page);
    WebKitDOMElement* body = WEBKIT_DOM_ELEMENT(webkit_dom_document_get_body (document));
    glong * const pos = user_data;
    
    if (body){
        *pos = webkit_dom_element_get_scroll_top(body);        
    } else {
      g_print("couldn't get body!\n");
    }
    
    g_print("store_scroll_position()\n");
    g_print("  scroll position: %d\n", *pos);
    
    return FALSE;
}
```

The restore function works as expected, but every time the store function is called, the ``g_print("couldn't get body!\n");`` gets executed. For this reason, the ``pos`` variable is always ``0`` when restore is called.

I'm not sure why this happens, but we are a step closer to the solution now.

**Comment by fabiocolacio** [comment]
Also, on my machine, for a single refresh of the webpage, ``store_scroll_position()`` gets executed 16 times, followed by 1 execution of ``restore_scroll_position()``. I am not surewhy ``store_scroll_position()`` is called so many times, but perhaps this is related.

**Comment by fabiocolacio** [comment]
Okay so the multiple calls to ``store_scroll_position()`` are due to files linked into the html document such as CSS and Javascripts, because webkit needs to request each of those resources separately.

**Comment by fabiocolacio** [comment]
Here is a sample output when refreshing once (with the document halfway scrolled):

```
store_scroll_position()
  scroll position: 568
store_scroll_position()
  couldn't get body!
  scroll position: 568
store_scroll_position()
  couldn't get body!
  scroll position: 568
store_scroll_position()
  couldn't get body!
  scroll position: 568
store_scroll_position()
  couldn't get body!
  scroll position: 568
store_scroll_position()
  couldn't get body!
  scroll position: 568
store_scroll_position()
  couldn't get body!
  scroll position: 568
store_scroll_position()
  couldn't get body!
  scroll position: 568
store_scroll_position()
  scroll position: 0
store_scroll_position()
  scroll position: 0
store_scroll_position()
  scroll position: 0
restore_scroll_position()
  scroll position: 0
```

**Comment by fabiocolacio** [comment]
I did another test with all CSS and javascripts disabled. In this case, there was only one ``store_scroll_position()`` for each ``restore_scroll_position()``, because the document was self-contained. In this case, the webextension ran with no issues. As shown in the output above, the scroll position is reset to ``0`` while requesting resources for some reason.

**Comment by Mandarancio** [comment]
Also in my case the request for the css and javascript were triggering the callback now I added a check on the resource uri and storing the position only when the uri is "file:///" (that correspond to our html).

Maybe solve your problem (on my config everything still work ;) )

**Comment by fabiocolacio** [comment]
Your solution did not work for me, but it gave me a very similar idea implemented in commit 7e3d8cad150fa8d16de6d449fb69fd7cce213869, and now it works.
